### PR TITLE
#3904 Inheritance support java client retrofit gson

### PR DESCRIPTION
**Pull request by konopski**
### PR checklist

- [x] Read the [contribution guildelines](https://github.com/swagger-api/swagger-codegen/blob/master/CONTRIBUTING.md).
- [x] Ran the shell/batch script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates)
- [x] Filed the PR against the correct branch: master for non-breaking changes and `2.3.0` branch for breaking (non-backward compatible) changes.

### Description of the PR

This PR provides java client support for model inheritance when using retrofit2 and gson. It is a bit raw and I am aware that this may need some work before merge, but it would be nice to get some feedback regarding eg. how to test it or document. Is it ok to add new dependency? 
There is number of issues opened that are related to the problem. I picked the most generic one #3904.
Currently I assume that subclass must have exactly same name as value of discriminator. This complies to specification: **Composition and Inheritance (Polymorphism)**
_The value of the chosen property has to be the friendly name given to the model under the definitions property._

### CODE REVIEW

**Comment by wing328** [comment]
Thanks for the PR but your commit (as shown in the Commits tab) is not linked to your Github account, which means this PR won't count as your contribution in https://github.com/swagger-api/swagger-codegen/graphs/contributors.

Let me know if you need help fixing it.

Ref: https://github.com/swagger-api/swagger-codegen/wiki/FAQ#how-can-i-update-commits-that-are-not-linked-to-my-github-account

**Comment by konopski** [comment]
Thanks, I reconfigured the email.

**Comment by konopski** [comment]
I provided some description of this change:
http://tech.zooplus.com/generating-java-model-with-inheritance-from-swagger-specification/

**Comment by wing328** [comment]
@konopski sorry for taking this long. I'll review on coming Sunday and let you know if I've any question, and thanks for writing a blog post to explain it.

**Comment by wing328** [comment]
We plan to release current master (2.2.3) this week so we'll review this after the release

**Comment by wing328** [comment]
Added `zooplus` to the list of companies using Swagger Codegen via https://github.com/swagger-api/swagger-codegen/commit/8a1b030ed16ff4e81ee4558a411c8224eb4b6d2e

**Comment by pgordon9** [comment]
Will be borrowing the work from this for Scala Inheritance support ASAP.

**Comment by konopski** [comment]
I have just incorporated latest changes from master.
As far as I can see dedicated scala client rather uses jackson, this PR does not support it. But I guess you can use generated gson based java model in scala as it is.

**Comment by konopski** [comment]
Fixed bug in generated gradle file. 
I am preparing similar enhancement for OkHttp.

**Comment by konopski** [comment]
Any feedback regarding the PR?

**Comment by wing328** [comment]
@konopski I'm deeply sorry for the delay. I'll review tomorrow and ping you if I need any help.

**Comment by wing328** [comment]
cc @cbornet @ePaul

**Comment by cbornet** [comment]
@konopski could you generate the retrofit samples ?

File: samples/client/petstore/java/retrofit2/src/main/java/io/swagger/client/JSON.java [hunk 1]
```diff
@@ -39,8 +46,25 @@
...
     private LocalDateTypeAdapter localDateTypeAdapter = new LocalDateTypeAdapter();
 
+    public static GsonBuilder createGson() {
+        GsonFireBuilder fireBuilder = new GsonFireBuilder()
+          .registerTypeSelector(Animal.class, new TypeSelector() {
+            @Override
+            public Class getClassForElement(JsonElement readElement) {
+                Map classByDiscriminatorValue = new HashMap();
+                classByDiscriminatorValue.put("Cat".toUpperCase(), Cat.class);
+                classByDiscriminatorValue.put("Dog".toUpperCase(), Dog.class);
```
Comment (cbornet) [inline samples/client/petstore/java/retrofit2/src/main/java/io/swagger/client/JSON.java]: Isn't `Animal` a valid discriminator value ?

File: samples/client/petstore/java/retrofit2/src/main/java/io/swagger/client/JSON.java [hunk 2]
```diff
@@ -39,8 +46,25 @@
...
+                String discriminatorField = "className";
+                String discriminatorValue = readElement.getAsJsonObject().get(discriminatorField).getAsString();
```
Comment (cbornet) [inline samples/client/petstore/java/retrofit2/src/main/java/io/swagger/client/JSON.java]: The case where `discriminatorField` is not in the JSON should be handled

File: samples/client/petstore/java/retrofit2/src/main/java/io/swagger/client/JSON.java [hunk 3]
```diff
@@ -39,8 +46,25 @@
...
+                return (Class) classByDiscriminatorValue.get(discriminatorValue.toUpperCase());
```
Comment (cbornet) [inline samples/client/petstore/java/retrofit2/src/main/java/io/swagger/client/JSON.java]: The case where `discriminatorValue` is not in `classByDiscriminatorValue` should be handled cleanly.

File: samples/client/petstore/java/retrofit2/src/main/java/io/swagger/client/JSON.java (hunk 1 above)
Comment (konopski) [reply samples/client/petstore/java/retrofit2/src/main/java/io/swagger/client/JSON.java]: I assumed my case with abstract superclass would be most typical. I added support for it and sample is including also `Animal`.

File: modules/swagger-codegen/src/main/java/io/swagger/codegen/DefaultGenerator.java [hunk 4]
```diff
@@ -736,6 +736,7 @@ public Reader getTemplate(String name) {
         return files;
     }
 
+
```
Comment (cbornet) [inline modules/swagger-codegen/src/main/java/io/swagger/codegen/DefaultGenerator.java]: Extra line

**Comment by cbornet** [comment]
LGTM

**Pull request merged**